File: packages/core/src/db/seed.ts
import { getConfig, isConfigInitialized } from '../config/index';
import type { AccountType } from '../types/ledger';

export type ChartAccountSeed = {
	id: string;
	name: string;
	type: AccountType;
	parent: string | null;
	placeholder: boolean;
};

/**
 * Generate asset account seeds from TOML config.
 * Creates placeholder parents and leaf accounts dynamically.
 */
function getAssetAccountSeeds(): ChartAccountSeed[] {
	if (!isConfigInitialized()) {
		// Fallback when config not loaded (shouldn't happen in normal operation)
		return [];
	}

	const accounts = getConfig().accounts.filter((a) => a.type === 'asset');
	const seeds: ChartAccountSeed[] = [];
	const addedParents = new Set<string>();

	// Root Assets placeholder
	seeds.push({ id: 'Assets', name: 'Assets', type: 'asset', parent: null, placeholder: true });
	addedParents.add('Assets');

	for (const account of accounts) {
		// Parse the account ID to get parent hierarchy
		// e.g., "Assets:Business:Wise" -> parent "Assets:Business"
		const parts = account.id.split(':');

		// Add intermediate placeholder parents
		for (let i = 1; i < parts.length; i++) {
			const parentId = parts.slice(0, i + 1).join(':');
			const grandParentId = i === 1 ? 'Assets' : parts.slice(0, i).join(':');

			if (!addedParents.has(parentId) && parentId !== account.id) {
				seeds.push({
					id: parentId,
					name: parts[i] ?? parentId,
					type: 'asset',
					parent: grandParentId,
					placeholder: true,
				});
				addedParents.add(parentId);
			}
		}

		// Add the leaf account
		const parentParts = parts.slice(0, -1);
		const parentId = parentParts.length > 0 ? parentParts.join(':') : null;
		const accountName = account.label ?? parts[parts.length - 1] ?? account.id;

		seeds.push({
			id: account.id,
			name: accountName,
			type: 'asset',
			parent: parentId,
			placeholder: false,
		});
	}

	return seeds;
}

// Static accounts that don't change based on config
const STATIC_SEEDS: ChartAccountSeed[] = [
	// Root accounts (non-asset placeholders)
	{ id: 'Liabilities', name: 'Liabilities', type: 'liability', parent: null, placeholder: true },
	{ id: 'Liabilities:Business', name: 'Business', type: 'liability', parent: 'Liabilities', placeholder: true },
	{ id: 'Liabilities:Business:CorpTaxPayable', name: 'Corp Tax Payable', type: 'liability', parent: 'Liabilities:Business', placeholder: false },
	{ id: 'Liabilities:Business:VATPayable', name: 'VAT Payable', type: 'liability', parent: 'Liabilities:Business', placeholder: false },
	{ id: 'Equity', name: 'Equity', type: 'equity', parent: null, placeholder: true },
	{ id: 'Income', name: 'Income', type: 'income', parent: null, placeholder: true },
	{ id: 'Expenses', name: 'Expenses', type: 'expense', parent: null, placeholder: true },

	// Equity
	{ id: 'Equity:OpeningBalances', name: 'Opening Balances', type: 'equity', parent: 'Equity', placeholder: false },
	{ id: 'Equity:RetainedEarnings', name: 'Retained Earnings', type: 'equity', parent: 'Equity', placeholder: false },
	{ id: 'Equity:Transfers', name: 'Internal Transfers', type: 'equity', parent: 'Equity', placeholder: false },

	// Income hierarchy
	{ id: 'Income:Salary', name: 'Salary', type: 'income', parent: 'Income', placeholder: false },
	{ id: 'Income:Dividends', name: 'Dividends', type: 'income', parent: 'Income', placeholder: false },
	{ id: 'Income:Interest', name: 'Interest', type: 'income', parent: 'Income', placeholder: false },
	{ id: 'Income:Refunds', name: 'Refunds', type: 'income', parent: 'Income', placeholder: false },
	{ id: 'Income:Other', name: 'Other', type: 'income', parent: 'Income', placeholder: false },

	// Expense hierarchy
	{ id: 'Expenses:Food', name: 'Food', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Food:Groceries', name: 'Groceries', type: 'expense', parent: 'Expenses:Food', placeholder: false },
	{ id: 'Expenses:Food:Restaurants', name: 'Restaurants', type: 'expense', parent: 'Expenses:Food', placeholder: false },
	{ id: 'Expenses:Food:Coffee', name: 'Coffee', type: 'expense', parent: 'Expenses:Food', placeholder: false },
	{ id: 'Expenses:Food:Delivery', name: 'Delivery', type: 'expense', parent: 'Expenses:Food', placeholder: false },
	{ id: 'Expenses:Food:Supplements', name: 'Supplements', type: 'expense', parent: 'Expenses:Food', placeholder: false },

	{ id: 'Expenses:Housing', name: 'Housing', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Housing:Rent', name: 'Rent', type: 'expense', parent: 'Expenses:Housing', placeholder: false },
	{ id: 'Expenses:Housing:Utilities', name: 'Utilities', type: 'expense', parent: 'Expenses:Housing', placeholder: false },
	{ id: 'Expenses:Housing:Insurance', name: 'Insurance', type: 'expense', parent: 'Expenses:Housing', placeholder: false },
	{ id: 'Expenses:Housing:Maintenance', name: 'Maintenance', type: 'expense', parent: 'Expenses:Housing', placeholder: false },

	{ id: 'Expenses:Transport', name: 'Transport', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Transport:PublicTransport', name: 'Public Transport', type: 'expense', parent: 'Expenses:Transport', placeholder: false },
	{ id: 'Expenses:Transport:Fuel', name: 'Fuel', type: 'expense', parent: 'Expenses:Transport', placeholder: false },
	{ id: 'Expenses:Transport:Parking', name: 'Parking', type: 'expense', parent: 'Expenses:Transport', placeholder: false },
	{ id: 'Expenses:Transport:Maintenance', name: 'Maintenance', type: 'expense', parent: 'Expenses:Transport', placeholder: false },
	{ id: 'Expenses:Transport:Taxi', name: 'Taxi', type: 'expense', parent: 'Expenses:Transport', placeholder: false },
	{ id: 'Expenses:Transport:Vehicle', name: 'Vehicle', type: 'expense', parent: 'Expenses:Transport', placeholder: false },

	{ id: 'Expenses:Business', name: 'Business', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Business:Software', name: 'Software', type: 'expense', parent: 'Expenses:Business', placeholder: false },
	{ id: 'Expenses:Business:Equipment', name: 'Equipment', type: 'expense', parent: 'Expenses:Business', placeholder: false },
	{ id: 'Expenses:Business:Services', name: 'Services', type: 'expense', parent: 'Expenses:Business', placeholder: false },
	{ id: 'Expenses:Business:BankFees', name: 'Bank Fees', type: 'expense', parent: 'Expenses:Business', placeholder: false },
	{ id: 'Expenses:Business:Legal', name: 'Legal', type: 'expense', parent: 'Expenses:Business', placeholder: false },
	{ id: 'Expenses:Business:Accounting', name: 'Accounting', type: 'expense', parent: 'Expenses:Business', placeholder: false },
	{ id: 'Expenses:Business:Insurance', name: 'Insurance', type: 'expense', parent: 'Expenses:Business', placeholder: false },

	{ id: 'Expenses:Entertainment', name: 'Entertainment', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Entertainment:Subscriptions', name: 'Subscriptions', type: 'expense', parent: 'Expenses:Entertainment', placeholder: false },
	{ id: 'Expenses:Entertainment:Leisure', name: 'Leisure', type: 'expense', parent: 'Expenses:Entertainment', placeholder: false },
	{ id: 'Expenses:Entertainment:Gaming', name: 'Gaming', type: 'expense', parent: 'Expenses:Entertainment', placeholder: false },

	{ id: 'Expenses:Health', name: 'Health', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Health:Medical', name: 'Medical', type: 'expense', parent: 'Expenses:Health', placeholder: false },
	{ id: 'Expenses:Health:Pharmacy', name: 'Pharmacy', type: 'expense', parent: 'Expenses:Health', placeholder: false },
	{ id: 'Expenses:Health:Fitness', name: 'Fitness', type: 'expense', parent: 'Expenses:Health', placeholder: false },
	{ id: 'Expenses:Health:Insurance', name: 'Insurance', type: 'expense', parent: 'Expenses:Health', placeholder: false },

	{ id: 'Expenses:Shopping', name: 'Shopping', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Shopping:Clothing', name: 'Clothing', type: 'expense', parent: 'Expenses:Shopping', placeholder: false },
	{ id: 'Expenses:Shopping:Electronics', name: 'Electronics', type: 'expense', parent: 'Expenses:Shopping', placeholder: false },
	{ id: 'Expenses:Shopping:Home', name: 'Home', type: 'expense', parent: 'Expenses:Shopping', placeholder: false },

	{ id: 'Expenses:Personal', name: 'Personal', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Personal:Gifts', name: 'Gifts', type: 'expense', parent: 'Expenses:Personal', placeholder: false },
	{ id: 'Expenses:Personal:Education', name: 'Education', type: 'expense', parent: 'Expenses:Personal', placeholder: false },
	{ id: 'Expenses:Personal:Charity', name: 'Charity', type: 'expense', parent: 'Expenses:Personal', placeholder: false },

	{ id: 'Expenses:Taxes', name: 'Taxes', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Taxes:IncomeTax', name: 'Income Tax', type: 'expense', parent: 'Expenses:Taxes', placeholder: false },
	{ id: 'Expenses:Taxes:NationalInsurance', name: 'National Insurance', type: 'expense', parent: 'Expenses:Taxes', placeholder: false },
	{ id: 'Expenses:Taxes:VAT', name: 'VAT', type: 'expense', parent: 'Expenses:Taxes', placeholder: false },

	{ id: 'Expenses:Bills', name: 'Bills', type: 'expense', parent: 'Expenses', placeholder: true },
	{ id: 'Expenses:Bills:Energy', name: 'Energy', type: 'expense', parent: 'Expenses:Bills', placeholder: false },
	{ id: 'Expenses:Bills:Water', name: 'Water', type: 'expense', parent: 'Expenses:Bills', placeholder: false },
	{ id: 'Expenses:Bills:CouncilTax', name: 'Council Tax', type: 'expense', parent: 'Expenses:Bills', placeholder: false },
	{ id: 'Expenses:Bills:Internet', name: 'Internet', type: 'expense', parent: 'Expenses:Bills', placeholder: false },
	{ id: 'Expenses:Bills:Insurance', name: 'Insurance', type: 'expense', parent: 'Expenses:Bills', placeholder: false },
	{ id: 'Expenses:Bills:DirectDebits', name: 'Direct Debits', type: 'expense', parent: 'Expenses:Bills', placeholder: false },

	{ id: 'Expenses:Uncategorized', name: 'Uncategorized', type: 'expense', parent: 'Expenses', placeholder: false },
];

/**
 * Get the complete chart of accounts seed data.
 * Combines dynamic asset accounts from config with static expense/income/equity accounts.
 */
export function getChartOfAccountsSeeds(): ChartAccountSeed[] {
	const assetSeeds = getAssetAccountSeeds();
	return [...assetSeeds, ...STATIC_SEEDS];
}

File: crates/fin-sdk/src/db/seed.rs
use std::collections::HashSet;

use rusqlite::{Connection, params};

use crate::config::FinConfig;
use crate::error::Result;

#[derive(Debug, Clone)]
struct ChartAccountSeed {
    id: String,
    name: String,
    account_type: String,
    parent_id: Option<String>,
    is_placeholder: bool,
}

fn static_seeds() -> Vec<ChartAccountSeed> {
    [
        ("Liabilities", "Liabilities", "liability", None, true),
        (
            "Liabilities:Business",
            "Business",
            "liability",
            Some("Liabilities"),
            true,
        ),
        (
            "Liabilities:Business:CorpTaxPayable",
            "Corp Tax Payable",
            "liability",
            Some("Liabilities:Business"),
            false,
        ),
        (
            "Liabilities:Business:VATPayable",
            "VAT Payable",
            "liability",
            Some("Liabilities:Business"),
            false,
        ),
        ("Equity", "Equity", "equity", None, true),
        (
            "Equity:OpeningBalances",
            "Opening Balances",
            "equity",
            Some("Equity"),
            false,
        ),
        (
            "Equity:RetainedEarnings",
            "Retained Earnings",
            "equity",
            Some("Equity"),
            false,
        ),
        (
            "Equity:Transfers",
            "Internal Transfers",
            "equity",
            Some("Equity"),
            false,
        ),
        (
            "Equity:Investments",
            "Investments",
            "equity",
            Some("Equity"),
            false,
        ),
        ("Income", "Income", "income", None, true),
        ("Income:Salary", "Salary", "income", Some("Income"), false),
        (
            "Income:Dividends",
            "Dividends",
            "income",
            Some("Income"),
            false,
        ),
        (
            "Income:Interest",
            "Interest",
            "income",
            Some("Income"),
            false,
        ),
        ("Income:Refunds", "Refunds", "income", Some("Income"), false),
        ("Income:Other", "Other", "income", Some("Income"), false),
        ("Expenses", "Expenses", "expense", None, true),
        ("Expenses:Food", "Food", "expense", Some("Expenses"), true),
        (
            "Expenses:Food:Groceries",
            "Groceries",
            "expense",
            Some("Expenses:Food"),
            false,
        ),
        (
            "Expenses:Food:Restaurants",
            "Restaurants",
            "expense",
            Some("Expenses:Food"),
            false,
        ),
        (
            "Expenses:Food:Supplements",
            "Supplements",
            "expense",
            Some("Expenses:Food"),
            false,
        ),
        (
            "Expenses:Transport",
            "Transport",
            "expense",
            Some("Expenses"),
            true,
        ),
        (
            "Expenses:Transport:PublicTransport",
            "Public Transport",
            "expense",
            Some("Expenses:Transport"),
            false,
        ),
        (
            "Expenses:Transport:Travel",
            "Travel",
            "expense",
            Some("Expenses:Transport"),
            false,
        ),
        (
            "Expenses:Transport:Vehicle",
            "Vehicle",
            "expense",
            Some("Expenses:Transport"),
            false,
        ),
        (
            "Expenses:Transport:Parking",
            "Parking",
            "expense",
            Some("Expenses:Transport"),
            false,
        ),
        (
            "Expenses:Housing",
            "Housing",
            "expense",
            Some("Expenses"),
            true,
        ),
        (
            "Expenses:Housing:Rent",
            "Rent",
            "expense",
            Some("Expenses:Housing"),
            false,
        ),
        (
            "Expenses:Housing:Utilities",
            "Utilities",
            "expense",
            Some("Expenses:Housing"),
            false,
        ),
        (
            "Expenses:Business",
            "Business",
            "expense",
            Some("Expenses"),
            true,
        ),
        (
            "Expenses:Business:Software",
            "Software",
            "expense",
            Some("Expenses:Business"),
            false,
        ),
        (
            "Expenses:Business:Equipment",
            "Equipment",
            "expense",
            Some("Expenses:Business"),
            false,
        ),
        (
            "Expenses:Business:Services",
            "Services",
            "expense",
            Some("Expenses:Business"),
            false,
        ),
        (
            "Expenses:Business:Contractors",
            "Contractors",
            "expense",
            Some("Expenses:Business"),
            false,
        ),
        (
            "Expenses:Business:Insurance",
            "Insurance",
            "expense",
            Some("Expenses:Business"),
            false,
        ),
        (
            "Expenses:Entertainment",
            "Entertainment",
            "expense",
            Some("Expenses"),
            true,
        ),
        (
            "Expenses:Entertainment:Subscriptions",
            "Subscriptions",
            "expense",
            Some("Expenses:Entertainment"),
            false,
        ),
        (
            "Expenses:Entertainment:Leisure",
            "Leisure",
            "expense",
            Some("Expenses:Entertainment"),
            false,
        ),
        (
            "Expenses:Health",
            "Health",
            "expense",
            Some("Expenses"),
            true,
        ),
        (
            "Expenses:Health:Medical",
            "Medical",
            "expense",
            Some("Expenses:Health"),
            false,
        ),
        (
            "Expenses:Health:Fitness",
            "Fitness",
            "expense",
            Some("Expenses:Health"),
            false,
        ),
        (
            "Expenses:Health:Insurance",
            "Insurance",
            "expense",
            Some("Expenses:Health"),
            false,
        ),
        (
            "Expenses:Shopping",
            "Shopping",
            "expense",
            Some("Expenses"),
            true,
        ),
        (
            "Expenses:Shopping:Home",
            "Home",
            "expense",
            Some("Expenses:Shopping"),
            false,
        ),
        (
            "Expenses:Shopping:Charity",
            "Charity",
            "expense",
            Some("Expenses:Shopping"),
            false,
        ),
        (
            "Expenses:Personal",
            "Personal",
            "expense",
            Some("Expenses"),
            true,
        ),
        (
            "Expenses:Personal:Immigration",
            "Immigration",
            "expense",
            Some("Expenses:Personal"),
            false,
        ),
        ("Expenses:Taxes", "Taxes", "expense", Some("Expenses"), true),
        (
            "Expenses:Taxes:VAT",
            "VAT",
            "expense",
            Some("Expenses:Taxes"),
            false,
        ),
        (
            "Expenses:Taxes:HMRC",
            "HMRC",
            "expense",
            Some("Expenses:Taxes"),
            false,
        ),
        (
            "Expenses:Taxes:CorporationTax",
            "Corporation Tax",
            "expense",
            Some("Expenses:Taxes"),
            false,
        ),
        (
            "Expenses:Taxes:SelfAssessment",
            "Self Assessment",
            "expense",
            Some("Expenses:Taxes"),
            false,
        ),
        (
            "Expenses:Taxes:PAYE",
            "PAYE",
            "expense",
            Some("Expenses:Taxes"),
            false,
        ),
        ("Expenses:Bills", "Bills", "expense", Some("Expenses"), true),
        (
            "Expenses:Bills:Energy",
            "Energy",
            "expense",
            Some("Expenses:Bills"),
            false,
        ),
        (
            "Expenses:Bills:Water",
            "Water",
            "expense",
            Some("Expenses:Bills"),
            false,
        ),
        (
            "Expenses:Bills:CouncilTax",
            "Council Tax",
            "expense",
            Some("Expenses:Bills"),
            false,
        ),
        (
            "Expenses:Bills:Internet",
            "Internet",
            "expense",
            Some("Expenses:Bills"),
            false,
        ),
        (
            "Expenses:Bills:Insurance",
            "Insurance",
            "expense",
            Some("Expenses:Bills"),
            false,
        ),
        (
            "Expenses:Bills:DirectDebits",
            "Direct Debits",
            "expense",
            Some("Expenses:Bills"),
            false,
        ),
        (
            "Expenses:Uncategorized",
            "Uncategorized",
            "expense",
            Some("Expenses"),
            false,
        ),
        (
            "Expenses:Other",
            "Other",
            "expense",
            Some("Expenses"),
            false,
        ),
    ]
    .into_iter()
    .map(
        |(id, name, account_type, parent_id, is_placeholder)| ChartAccountSeed {
            id: id.to_owned(),
            name: name.to_owned(),
            account_type: account_type.to_owned(),
            parent_id: parent_id.map(std::string::ToString::to_string),
            is_placeholder,
        },
    )
    .collect()
}

fn dynamic_asset_seeds(config: &FinConfig) -> Vec<ChartAccountSeed> {
    let mut seeds = vec![ChartAccountSeed {
        id: "Assets".to_owned(),
        name: "Assets".to_owned(),
        account_type: "asset".to_owned(),
        parent_id: None,
        is_placeholder: true,
    }];
    let mut seen = HashSet::from(["Assets".to_owned()]);

    for account in config
        .accounts
        .iter()
        .filter(|account| account.account_type == "asset")
    {
        let parts = account.id.split(':').collect::<Vec<_>>();
        for index in 1..parts.len() {
            let parent_id = parts[0..=index].join(":");
            if parent_id == account.id || seen.contains(&parent_id) {
                continue;
            }
            let grand_parent = if index == 1 {
                Some("Assets".to_owned())
            } else {
                Some(parts[0..index].join(":"))
            };
            let name = parts[index].to_owned();
            seeds.push(ChartAccountSeed {
                id: parent_id.clone(),
                name,
                account_type: "asset".to_owned(),
                parent_id: grand_parent,
                is_placeholder: true,
            });
            seen.insert(parent_id);
        }

        let leaf_name = account
            .label
            .clone()
            .unwrap_or_else(|| parts.last().unwrap_or(&account.id.as_str()).to_string());
        let parent_id = if parts.len() > 1 {
            Some(parts[0..parts.len() - 1].join(":"))
        } else {
            None
        };
        seeds.push(ChartAccountSeed {
            id: account.id.clone(),
            name: leaf_name,
            account_type: "asset".to_owned(),
            parent_id,
            is_placeholder: false,
        });
    }

    seeds
}

pub fn ensure_chart_of_accounts_seeded(connection: &Connection, config: &FinConfig) -> Result<()> {
    let mut seeds = dynamic_asset_seeds(config);
    seeds.extend(static_seeds());

    let mut statement = connection.prepare(
        "INSERT OR IGNORE INTO chart_of_accounts (id, name, account_type, parent_id, is_placeholder)\n         VALUES (?1, ?2, ?3, ?4, ?5)",
    )?;

    for seed in seeds {
        statement.execute(params![
            seed.id,
            seed.name,
            seed.account_type,
            seed.parent_id,
            i32::from(seed.is_placeholder),
        ])?;
    }
    Ok(())
}
